### PR TITLE
Create Model for tracking NLP usage

### PR DESCRIPTION
**Pull request by JacquelineMorrissette**
## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a description of your work suitable for publishing on [our forum](https://community.kobotoolbox.org/tag/release-notes)
6. [x] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

Created a model to track the usage of NLP services (Automatic Speech Recognition and Machine Translation) by service provider

## Related issues
part of #3098

### CODE REVIEW

File: kobo/apps/trackers/utils.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    sql = f"""
+        jsonb_set(
+            counters,
+            '{{{service}}}',
+            (COALESCE(counters->>'{service}','0')::int + {amount})::text::jsonb
+        )
+    """
+    MonthlyNLPUsageCounter.objects.filter(**criteria).update(
+        counters=RawSQL(sql, [])
+    )
```
Comment (noliveleger) [inline kobo/apps/trackers/utils.py:47]: What about adding a new `Func` child class in `kpi/utils/jsonbfield_helper.py`

```python 
class IncrementValue(Func):

    function = 'jsonb_set'
    template = (
        "%(function)s(%(expressions)s,"
        "'{\"%(keyname)s\"}',"
        "(COALESCE(%(expressions)s ->> '%(keyname)s', '0')::int "
        "+ %(increment)s)::text::jsonb)"
    )
    arity = 1

    def __init__(self, expression: str, keyname: str, increment: int, **extra):
        super().__init__(
            expression,
            keyname=keyname,
            increment=increment,
            **extra,
        )

        
class ReplaceValues(Func):
    ...
```

and use it in your  queryset this way: 
```python 
MonthlyNLPUsageCounter.objects.filter(**criteria).update(
    counters=IncrementValue('counters', keyname=service, increment=amount)
)
```

File: kobo/apps/trackers/models.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            sql = f"""
+                jsonb_set(
+                    counters,
+                    '{{{service}}}',
+                    (COALESCE(counters->>'{service}','0')::int + {instance.counters[service]})::text::jsonb
+                )
+            """
+            MonthlyNLPUsageCounter.objects.filter(**criteria).update(
+                counters=RawSQL(sql, [])
+            )
```
Comment (noliveleger) [inline kobo/apps/trackers/models.py:54]: What about calling your utility function directly? 

```python 
def update_catch_all_counter_on_delete(cls, sender, instance, **kwargs):
     criteria = dict(
           year=instance.year,
           month=instance.month,
           user_id=instance.user_id,
           asset=None,
     )
     # make sure an instance exists with `asset = NULL`
     counter = cls.objects.get_or_create(**criteria)

    services = instance.counters.keys()
    for service in instances.counters.keys():
        update_nlp_counter(service, amount=instances.counters[service], user_id=instance.user_id, counter_id=counter.pk)
...

def update_nlp_counter(
        service: str,
        amount: int,
        user_id: int,
        asset_id: Optional[int] = None,
        counter_id: Optional[int] = None,
):
    if not counter_id:
        date = datetime.today()
        criteria = dict(
            year=date.year,
            month=date.month,
            user_id=user_id,
            asset_id=asset_id,
       )
      # Ensure the counter for the month exists first
      counter =  MonthlyNLPUsageCounter.objects.get_or_create(**criteria)
      counter_id = counter.pk
      
      MonthlyNLPUsageCounter.objects.filter(pk=counter_id).update(
          counters=IncrementValue('counters', keyname=service, increment=amount)
      )
```

File: kobo/apps/trackers/models.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            UniqueConstraint(fields=['year', 'month', 'user'],
+                             condition=Q(asset=None),
+                             name='unique_without_asset')
+        ]
+        indexes = [
+            models.Index(fields=('year', 'month', 'user')),
+        ]
+
+    @classmethod
+    def update_catch_all_counter_on_delete(cls, sender, instance, **kwargs):
```
Comment (noliveleger) [inline kobo/apps/trackers/models.py:30]: Should it be `update_catch_all_counters_on_delete`?  (plural)

File: kobo/apps/trackers/tests.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        asset.delete()
+        assert MonthlyNLPUsageCounter.objects.all().count() == 1
+        tracker_no_asset = MonthlyNLPUsageCounter.objects.filter(
+            user=self.user,
+            asset_id=None,
+        )
+        tracker_with_asset = MonthlyNLPUsageCounter.objects.filter(
+            asset_id=a_id,
+        )
+        assert tracker_no_asset.count() == 1
```
Comment (noliveleger) [inline kobo/apps/trackers/tests.py:57]: Maybe we should assert it equals 0 before calling `asset.delete()` . 
`assert MonthlyNLPUsageCounter.objects.all().count() == 1` line 49 does not make it obvious that the existing one if for `asset = None`.

File: kobo/apps/trackers/tests.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        )
+        assert tracker_updated_service_amount.counters[service] == expected_amount
+
+        # ensure original tracker stays with new service added
+        update_nlp_counter(new_service, initial_amount, self.user.id, asset.id)
+        tracker_two_services = MonthlyNLPUsageCounter.objects.get(
+            user_id=self.user.id,
+            asset_id=asset.id,
+        )
+        assert tracker_two_services.counters[new_service] == initial_amount
```
Comment (noliveleger) [inline kobo/apps/trackers/tests.py:92]: Maybe test that service does not exist before calling `update_nlp_counter(new_service)` or it equals 0.

File: kobo/apps/trackers/tests.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import datetime
```
Comment (noliveleger) [inline kobo/apps/trackers/tests.py:1]: You can import `from datetime import datetime` to avoid calling `datetime.datetime` everywhere.

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kobo/apps/trackers/models.py (hunk 2 above)
Comment (JacquelineMorrissette) [reply kobo/apps/trackers/models.py:54]: I avoided calling the utility function here because it caused a circular import

Comment (noliveleger) [reply kobo/apps/trackers/models.py:54]: @JacquelineMorrissette, let's talk about this.

File: kobo/apps/trackers/utils.py [hunk 7]
```diff
@@ -1,7 +1,9 @@
 from datetime import datetime
 from typing import Optional
 
-from kobo.apps.trackers.models import MonthlyNLPUsageCounter
+# from kobo.apps.trackers.models import MonthlyNLPUsageCounter
```
Comment (noliveleger) [inline kobo/apps/trackers/utils.py:4]: My bad! Can you remove the commented import 🙏?

**Review by noliveleger — APPROVED** [review]

**Review by noliveleger — APPROVED** [review]

**Pull request merged**